druid-shell/src/platform/x11/window.rs

```diff
@@ -0,0 +1,444 @@
+        );
+
+        conn.flush();
+
+        let mut handler = self.handler.unwrap();
+        let handle = WindowHandle::new(window_id);
+        handler.connect(&(handle.clone()).into());
+
+        let xwindow = XWindow::new(window_id, handler, self.size);
+        Application::add_xwindow(window_id, xwindow);
+
+        Ok(handle)
+    }
+}
+
+// X11-specific event handling and window drawing (etc.)
+pub(crate) struct XWindow {
+    window_id: u32,
+    handler: Box<dyn WinHandler>,
+    refresh_rate: Option<f64>,
+}
+
+impl XWindow {
+    pub fn new(window_id: u32, window_handler: Box<dyn WinHandler>, size: Size) -> XWindow {
+        let conn = Application::get_connection();
+        let setup = conn.get_setup();
+        let screen_num = Application::get_screen_num();
+        // TODO(x11/errors): Don't unwrap for screen or visual_type?
+        let screen = setup.roots().nth(screen_num as usize).unwrap();
+        let mut visual_type = get_visual_from_screen(&screen).unwrap();
+
+        // Figure out the refresh rate of the current screen
+        let refresh_rate = util::refresh_rate(&conn, window_id);
+
+        let mut xwindow = XWindow {
+            window_id,
+            handler: window_handler,
+            refresh_rate,
+        };
+        // Let the window handler know the size of the window
+        xwindow.communicate_size(size);
+
+        xwindow
+    }
+
+    pub fn render(&mut self) {
+        let conn = Application::get_connection();
+        let setup = conn.get_setup();
+        let screen_num = Application::get_screen_num();
+        // TODO(x11/errors): Don't unwrap for screen or visual_type?
+        let screen = setup.roots().nth(screen_num as usize).unwrap();
+        let mut visual_type = get_visual_from_screen(&screen).unwrap();
+
+        // Figure out the window's current size
+        let geometry_cookie = xcb::get_geometry(&conn, self.window_id);
+        let reply = geometry_cookie.get_reply().unwrap();
+        let size = Size::new(reply.width() as f64, reply.height() as f64);
+        self.communicate_size(size);
+
+        // Create a draw surface
+        // TODO(x11/render_improvements): We have to re-create this draw surface if the window size changes.
+        let cairo_xcb_connection = unsafe {
+            cairo::XCBConnection::from_raw_none(
+                conn.get_raw_conn() as *mut cairo_sys::xcb_connection_t
+            )
+        };
+        let cairo_drawable = cairo::XCBDrawable(self.window_id);
+        let cairo_visual_type = unsafe {
+            cairo::XCBVisualType::from_raw_none(
+                &mut visual_type.base as *mut _ as *mut cairo_sys::xcb_visualtype_t,
+            )
+        };
+        let cairo_surface = cairo::XCBSurface::create(
+            &cairo_xcb_connection,
+            &cairo_drawable,
+            &cairo_visual_type,
+            size.width as i32,
+            size.height as i32,
+        )
+            .expect("couldn't create a cairo surface");
```

Comment:
This and one line in `application.rs` needs a `cargo fmt`

Reply:
Done